### DISCUSSION TITLE
PEP 723 clues

### DISCUSSION THREAD

**Discussion by alexpdp7** [pypa/pipx#1162] [Q&A]
Hi,

I have a fairly bizarre use case...

I have managed to set up my environment to exert PEP 723 support, following [this test](https://github.com/pypa/pipx/blob/1a3e6adf651d57b4750d7e6cfd5dca90c5080a29/tests/test_run.py#L195); I can put that script in a file, and get it to run using `pipx run test.py`.

But I would like to create script which has a Python module in the same directory as a dependency.

I'm developing a set of tools to use as a dependency for scripts you create using PEP 723. So you would do:

```
# /// pyproject
# run.requirements = ["mything"]
# ///

mything.foo()
```

However, the tool itself would have some of those scripts. This is an edge case, but I would like to get it working. I have tried variants like:

edit: I mean, instead of fetching `mything` from pypi, or to get it through git+https or whatever, I'd like to use the `mything` source which is just "right there" in `.`.

```
# /// pyproject
# run.requirements = ["mything @ ."]
# ///
```

or

```
# /// pyproject
# run.requirements = ["."]
# ///
```

, but they are not working for me. Is there a way to get it to work?

New stuff:

OK, this is strange:

```
$ cat scripts.py 
#!/usr/bin/env -S pipx run
# /// pyproject
# run.requirements = ["foo @ file://"]
# ///
import spl


spl.foo()
```

```
$ ./scripts.py 
Fatal error from pip prevented installation. Full pip output in file:
    /.../.local/pipx/logs/cmd_2023-12-22_18.55.20_pip_errors.log

Some possibly relevant errors from pip install:
    ERROR: Could not install packages due to an OSError: [Errno 21] Is a directory: '/.../git/scripts-py-libs'
Error installing foo@ file://.
```

```
$ ./scripts.py 
hi
```

I guess this manifests [the problem of using a flat layout instead of a src layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/)? But the installation error feels... like I'm getting closer?

Additional note:

In case anyone's wondering, this is what I'm doing <https://github.com/alexpdp7/scripts-py-libs>.

Cheers,

Álex

**Comment by alextremblay** [comment]
if you need to depend on external packages from pypi, you would want to use this pep-723-compliant syntax:

`scripts.py:`
```python
#!/usr/bin/env -S pipx run --path

# /// script
# requires-python = ">=3.11"
# dependencies = [
#   "requests",  # example of external dependency
# ]
# ///
import requests

requests.get(...)
```

but if your only dependency is this local spl package, you'll have a much easier time just telling python how to import it directly:

`scripts.py:`
```python
#!/usr/bin/env -S pipx run --path

import os
import sys

sys.path.append(os.path.dirname(__file__))  # tell python to import packages like spl from the folder that scripts.py live in

import spl

spl.do_stuff()
```

then your project layout would simply be:

```console
scripts.py
spl/__init__.py
spl/module_one.py
spl/module_two.py
```
... and so on

nice and simple :D

**Comment by alextremblay** [comment]
as an addendum, using the following systax...
```python
#!/usr/bin/env -S pipx run --path

# /// script
# dependencies = ["mything @ ./mything"]
# ///
```
... *would* work, if `mything` was a full-fledged python package, with its own pyproject.toml file, something like:

```console
scripts.py
mything/pyproject.toml
mything/src/mything/__init__.py
mything/src/mything/module_one.py
mything/src/mything/module_two.py
```

**Reply by alexpdp7** [reply]
Hi, thanks for the help!

I'm trying to do something pretty similar to your addendum (but with `mything` directly on `./`). But I can't get your example to work:

https://github.com/alexpdp7/pep723-pipx-poetry-experiments

This would be the "odd" case. Once I get that working, my "normal" use case would be something like:

```
#!/usr/bin/env -S pipx run --path

# /// script
# requires-python = ">=3.11"
# dependencies = ["git+https://github.com/alexpdp7/pep723-pipx-poetry-experiments.git"]
# ///

import pep723_pipx_poetry_experiments

pep723_pipx_poetry_experiments.foo()
```

Cheers,

Álex
